lib/jobs/ssh-job.js

```diff
@@ -0,0 +1,207 @@
+        assert.string(this.context.target);
+        this.nodeId = this.context.target;
+        this.acceptedCodes = (options.acceptedCodes || []).concat(0);
+        this.commands = this.buildCommands(options.commands);
+        assert.arrayOfObject(this.commands);
+    }
+    util.inherits(SshJob, BaseJob);
+
+    SshJob.prototype._run = function run() {
+        var self = this;
+        return waterline.nodes.needByIdentifier(self.nodeId)
+        .then(function(node) {
+            return Promise.reduce(self.commands, function(results, commandData) {
+                return self.sshExec(commandData.command, node.sshSettings, new Client())
+                .then(function(result) {
+                    result.catalogOptions = commandData.catalogOptions;
+                    result.cmd = commandData.command; //for parser matching
+                    return results.concat([result]);
+                });
+            }, []);
+        })
+        .then(self.handleResponse.bind(self))
+        .then(function() {
+            self._done();
+        })
+        .catch(self._done.bind(self));
+    };
+
+    SshJob.prototype.handleResponse = function(results) {
+        var self = this;
+
+        logger.debug("Received remote command output from node.", {
+            id: self.nodeId,
+        });
+
+        var catalogTasks = _.filter(results, function(result) {
+            logger.debug('task result', {data:result});
+            return  _.has(result, 'catalogOptions') && result.catalogOptions;
+        });
+
+        return self.catalogUserTasks(catalogTasks)
+        .catch(function(err) {
+            logger.error("Job error processing catalog output.", {
+                error: err,
+                id: self.nodeId,
+                taskContext: self.context
+            });
+            throw err;
+        });
+    };
+
+    SshJob.prototype.catalogUserTasks = function(tasks) {
+        var self = this;
+
+        return parser.parseUnknownTasks(tasks)
+        .spread(function() {
+            return Promise.map(Array.prototype.slice.call(arguments), function(result) {
+                if (result.error) {
+                    logger.error("Failed to parse data for " +
+                        result.source + ', ' + result.error,
+                        { error: result });
+                } else if (result.store) {
+                    return waterline.catalogs.create({
+                        node: self.nodeId,
+                        source: result.source || 'unknown',
+                        data: result.data
+                    });
+                } else {
+                    logger.info("Catalog result for " + result.source +
+                        " has not been marked as significant. Not storing.");
+                }
+            });
+        });
+    };
+
+    SshJob.prototype.sshExec = function(command, sshSettings, sshClient) {
+        var self = this;
+        return new Promise(function(resolve, reject) {
```

Comment:
I suggest adding a timeout to each SSH executive, so that each command will be finished within controlled time window.